### PR TITLE
[IMP] assets: support optimized rjsmin JS minification

### PR DESCRIPTION
**Pull request by odony**
The C implementation of the JS minification gives speedups between 6 and 55 times faster than the regex-based Python port, depending on how compressed the input it (which is what our default implementation does).
    
This is measurable when generating compiled assets bundle from scratch, e.g. after installing/updating modules or source code.
    
As an illustration, the minification of a 2MB JS bundle can be 50x faster:

```py
import rjsmin
from odoo.addons.base.models.assetsbundle import regex_jsmin
js_source = open("web.assets_common_lazy.js").read() # 2MB un-minified JS
%timeit regex_jsmin(js_source)
# -> 339 ms ± 495 µs per loop (mean ± std. dev. of 7 runs, 1 loop each)
%timeit rjsmin.jsmin(js_source)
# -> 6.79 ms ± 108 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
```

It's also a drop-in replacement, as long as rjsmin 1.1.0 or better is available (to support format strings properly, a.o.).
    
See also the documentation of rjsmin: http://opensource.perlig.de/rjsmin/

--

To discuss: remove the regex version entirely, to avoid any divergence of version/features between the two?

--
Update: let's drop the regex entirely for the sake of simplicity and avoiding duplication. The main argument for keeping it was to avoid bothering all developers with an update of their dev environments, but they're also likely to benefit from the speedup during developement and testing, so it's probably worth the temporary disruption.

### CODE REVIEW

**Comment by odony** [comment]
As discussed with @Xavier-Do and @antonylesuisse, a remaining question is: do we entirely drop the regex version in master, to avoid any future divergence? Or do we keep it for a while, to avoid armies of developers suddenly getting an error at startup?

**Comment by odony** [comment]
a first backport for saas-15.3 is in #104298, but this seems a possible candidate for backport in 15 and 16 too, I think

**Review by d-fence — APPROVED** [review]
LGTM from a packaging point of view.
To answer the removal question question, I would prefer to remove the regex version in master for the sake of keeping it simple.

File: odoo/addons/base/models/assetsbundle.py [hunk 1]
```diff
@@ -99,6 +99,12 @@ def subber(match):
     ).strip()
     return result
 
+try:
+    import rjsmin as rjsminlib
+    rjsmin = rjsminlib.jsmin
+except ImportError:
+    rjsmin = regex_jsmin
```
Comment (xmo-odoo) [inline odoo/addons/base/models/assetsbundle.py:106]:
```suggestion
try:
    from rjsmin import rjsmin
except ImportError:
    pass
```
and revert the name change on the function. Could also use `contextlib.suppress`, but that might require an additional import.

File: odoo/addons/base/models/assetsbundle.py [hunk 2]
```diff
@@ -99,6 +99,12 @@ def subber(match):
     ).strip()
     return result
 
+try:
+    import rjsmin as rjsminlib
+    rjsmin = rjsminlib.jsmin
+except ImportError:
```
Comment (Xavier-Do) [inline odoo/addons/base/models/assetsbundle.py:105]: If we don't remove the regex, I would at least add a warning that the lib is missing.

**Comment by Julien00859** [comment]
up @odoo/rd-framework-py

**Comment by Xavier-Do** [comment]
@robodoo r+

**Comment by tde-banana-odoo** [comment]
> Update: alright, let's drop the regex entirely for the sake of simplicity and avoiding duplication. The main argument for keeping it was to avoid bothering all developers with an update of their dev environments, but they're also likely to benefit from the speedup during developement and testing, so it's probably worth the temporary disruption. (pardon the typos)

Leave the devs alone !

**Comment by Xavier-Do** [comment]
> > Update: alright, let's drop the regex entirely for the sake of simplicity and avoiding duplication. The main argument for keeping it was to avoid bothering all developers with an update of their dev environments, but they're also likely to benefit from the speedup during developement and testing, so it's probably worth the temporary disruption. (pardon the typos)
> 
> Leave the devs alone !
You can already `sudo apt-get install python3-rjsmin` :laughing:

**Pull request closed without merging**